### PR TITLE
Text Classification with FNet [KerasNLP]

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves https://github.com/keras-team/keras-nlp/issues/213

Dataset: IMDb
Compared results with Transformer model

### CODE REVIEW

File: guides/keras_nlp/fnet_classification.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+2. The label is a sentiment score (float) lying in the range [0, 1]. We will
+convert it to either 0 or 1, keeping 0.5 as the threshold. This is a necessary
+step because the task we want to solve is a classification task.
+"""
+
+
+def preprocess_dataset(dataset):
+    dataset = dataset.map(
+        lambda x: {
+            "sentence": tf.strings.lower(x["sentence"]),
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:113]: Can't we lowercase inside the tokenizer? Why do that here?

File: guides/keras_nlp/fnet_classification.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+We'll be using `WordPieceTokenizer` from KerasNLP to tokenize the text.
+`WordPieceTokenizer` takes a WordPiece vocabulary and has functions for
+tokenizing the text, and detokenizing sequences of tokens.
+
+
+Before we define the tokenizer, we first need to train it on the dataset
+we have. WordPiece Tokenizer is a subword tokenizer; training it on a corpus gives
+us a vocabulary of subwords. A subword tokenizer is a compromise between word tokenizers
+(word tokenizers have the issue of many OOV tokens), and character tokenizers
+(characters don't really encode meaning like words do). Luckily, TensorFlow Text makes it very
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:147]: TensorFlow Text makes it very simple to train WordPiece on a corpus, as described in this [guide](link location)

File: guides/keras_nlp/fnet_classification.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+def train_word_piece(text_samples, vocab_size, reserved_tokens):
+    bert_tokenizer_params = dict(lower_case=True)
+
+    bert_vocab_args = dict(
+        # The target vocabulary size
+        vocab_size=vocab_size,
+        # Reserved tokens that must be included in the vocabulary
+        reserved_tokens=reserved_tokens,
+        # Arguments for `text.BertTokenizer`
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:165]: Same comments as other guide... `bert_tokenizer_params={"lower_case"=True}` remove `Arguments for `text.BertTokenizer` comment, remove `learn_params`.

File: guides/keras_nlp/fnet_classification.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    print("Sentence: ", input_sentence_ex)
+    print("Tokens: ", input_tokens_ex)
+    print("Recovered text after detokenizing: ", tokenizer.detokenize(input_tokens_ex))
+
+
+"""
+## Formatting the Dataset
+
+Next, we'll format our datasets in the form that will be fed to the models.
+We need to add [START] and [END] tokens to the input sentences. We also need
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:229]: I don't believe you need [START] and [END], and you aren't using them. Please remove.

File: guides/keras_nlp/fnet_classification.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+    sequence_length=MAX_SEQUENCE_LENGTH,
+)
+
+"""
+Let's try and tokenize a sample from our dataset! To verify whether the text has
+been tokenized correctly, we can also detokenize the list of tokens back to the
+original text.
+"""
+
+for element in train_ds.take(1):
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:216]: element = train_ds.take(1).get_single_element()

File: guides/keras_nlp/fnet_classification.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+collection of movie reviews labelled either positive or negative (sentiment
+analysis).
+
+To build the tokenizer, model, etc., we will use components from
+[KerasNLP](https://github.com/keras-team/keras-nlp). KerasNLP makes life easier
+for people who want to build NLP pipelines! :)
+
+### Model
+
+In 2017, a paper titled [Attention is All You Need](https://arxiv.org/abs/1706.03762)
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:24]: I would generally tighten up this section. Examples shouldn't have a ton of offhand comments, we should focus on what is show in this guide. This reads a little too much like a blog currently.

Roughly, we should just say...

 - BERT, RoBERTa, etc have shown the effectiveness of using transformers to compute a rich embedding for input text.
 - However transformers are expensive, an ongoing question is how to lower the compute requirements.
 - In this guide we will focus on FNet, which replace the expensive attention mechanism with a Fourier transform.
 - We will show how this can speed up training, without significantly degrading performance.

File: guides/keras_nlp/fnet_classification.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+Let's calculate the test accuracy.
+"""
+transformer_classifier.evaluate(test_ds, batch_size=BATCH_SIZE)
+
+"""
+We obtain a test accuracy of 73.62%.
+"""
+
+"""
+Let's make a table and compare the two models.
```
Comment (mattdangerw) [inline guides/keras_nlp/fnet_classification.py:371]: I would state this point a little more clearly--We can see that FNet significantly speeds up our run time, with only a small sacrifice in overall accuracy.

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks! Looks good. Left some initial comments.

File: guides/keras_nlp/fnet_classification.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+EMBED_DIM = 128
+INTERMEDIATE_DIM = 512
+
+"""
+## Loading the Dataset
+Tensorflow Datasets (TFDS) is a library that provides a unified API for working
+with datasets that are stored in the TensorFlow format. We will use TFDS to load
+the SST-2 dataset.
+"""
+train_ds, val_ds, test_ds = tfds.load(
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply guides/keras_nlp/fnet_classification.py:96]: One more thing - let's be more concise in the description, if we choose to use tfds, just say load SST-2 from Tensorflow Datasets (TFDS).

**Review by chenmoneygithub — COMMENTED** [review]
Thanks for the PR! left some comments

File: guides/keras_nlp/fnet_classification.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply guides/keras_nlp/fnet_classification.py:96]: This was a request from @fchollet when I was doing my guide, so maybe we should discuss with him? Personally not particularly opinionated.

[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply guides/keras_nlp/fnet_classification.py:96]: Ah okay, let's bring this up in the team chat. I feel since tfds is part of TF ecosystem and still being maintained, we should try using their product.

File: guides/keras_nlp/fnet_classification.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+[FNet: Mixing Tokens with Fourier Transforms](https://arxiv.org/abs/2105.03824)
+replaced the self-attention layer in BERT with a Fourier Transform layer. The
+magnitude of this was not lost on the research community; replacing a relatively
+complicated self-attention layer with a simpler "token mixing" layer resulted in
+comparable accuracy and a speed-up during training. This opens up further avenues
+of research - can we replace the self-attention layer with any simple "token mixing"
+layer and get comparable results?
+
+A couple of points from the paper stood out:
+1. The authors claim that FNet is 80% faster than BERT on GPUs and 70% faster on TPUs.
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py]: Hey,  the SST-2 dataset has very short sequences. I tried it with the IMDb dataset (which has longer sequences) and I'm getting a noticeable speed-up :D

File: guides/keras_nlp/fnet_classification.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+"""
+## Setup
+
+Before we start with the implementation, let's install the KerasNLP library, and
+import all the necessary packages.
+"""
+
+"""shell
+pip install -q keras-nlp
+pip install -q tfds-nightly
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py]: I used nightly because it has the `huggingface:sst` dataset. I've removed it now because I am using the IMDb dataset.

[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py]: I used nightly because it has the `huggingface:sst` dataset. I've removed it now because I am using the IMDb dataset.

File: guides/keras_nlp/fnet_classification.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py]: Removed TFDS. We don't need it for IMDb :)

File: guides/keras_nlp/fnet_classification.py (hunk 1 above)
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py]: It lowercases the special tokens as well, which I wanted to avoid.

File: guides/keras_nlp/fnet_classification.py [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+train_ds = make_dataset(train_ds)
+val_ds = make_dataset(val_ds)
+test_ds = make_dataset(test_ds)
+
+"""
+## Building the Model
+
+Now, let's move on to the exciting part - defining our model!
+We first need an embedding layer, i.e., a vector for every token in our input sequence.
```
Comment (chenmoneygithub) [inline guides/keras_nlp/fnet_classification.py:239]: a layer that maps every token in input sequence to a vector.

File: guides/keras_nlp/fnet_classification.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
+"""
+Title: Text Classification using FNet
+Author: [Abheesht Sharma](https://github.com/abheesht17/)
+Date created: 2021/06/01
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py:4]: Ah, man. Looks like I'm still mentally stuck in 2021 😆 . Changed!

File: guides/keras_nlp/fnet_classification.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+Samples are present in the form of text files. The directory contains two
+sub-directories: `train` and `test`. Each subdirectory in turn contains two
+folders: `pos` and `neg` for positive and negative reviews, respectively. We'll
+use the `tf.keras.utils.text_dataset_from_directory` utility to generate our
+labelled `tf.data.Dataset` dataset from text files. Before that, let's delete
+the unnecessary folders.
+"""
+
+"""shell
+ls aclImdb
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py:98]: Hmmm, I'll have to check this by generating the `.ipynb` file. Ideally, it should print all. I'll generate the `.iypnb` file and let you know.

**Review by abheesht17 — COMMENTED** [review]
Thanks for the review, @chenmoneygithub!

File: guides/keras_nlp/fnet_classification.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply guides/keras_nlp/fnet_classification.py:98]: None of this will get outputted in the rendered example the way things work on keras.io. If you want to do that you would need to actually os.listdir or something

[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py:98]: Ah, I see. Will change.

[the remark this answers is not in the document]
Comment (abheesht17) [reply guides/keras_nlp/fnet_classification.py:98]: Ah, I see. Will change.

File: guides/keras_nlp/fnet_classification.py [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+"""
+import keras_nlp
+import random
+import tensorflow as tf
+import os
+
+from tensorflow import keras
+from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab
+
+random.seed(42)
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:65]: Don't do this, instead use `keras.utils.set_random_seed()`

File: guides/keras_nlp/fnet_classification.py [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+"""shell
+pip install -q keras-nlp
+"""
+import keras_nlp
+import random
+import tensorflow as tf
+import os
+
+from tensorflow import keras
+from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:63]: Same as in the other example -- this should be hidden away

File: guides/keras_nlp/fnet_classification.py [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+"""shell
+rm -rf aclImdb/train/unsup
+"""
+
+"""
+We'll use the `tf.keras.utils.text_dataset_from_directory` utility to generate
+our labelled `tf.data.Dataset` dataset from text files.
+"""
+
+train_ds = tf.keras.utils.text_dataset_from_directory(
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:114]: Use `from tensorflow import keras` so you don't have to use `tf.keras` everywhere

File: guides/keras_nlp/fnet_classification.py [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+"""
+for text_batch, label_batch in train_ds.take(1):
+    for i in range(3):
+        print(text_batch.numpy()[i])
+        print(label_batch.numpy()[i])
+
+
+"""
+### Tokenizing the Data
+We'll be using the `keras_nlp.tokenizers.WordPieceTokenizer` layer to tokenize
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:150]: Add line break above

File: guides/keras_nlp/fnet_classification.py [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+    word_piece_ds = ds.unbatch().map(lambda x, y: x)
+    vocab = bert_vocab.bert_vocab_from_dataset(
+        word_piece_ds.batch(1000).prefetch(2), **bert_vocab_args
+    )
+    return vocab
+
+
+"""
+Every vocabulary has a few special, reserved tokens. We have four such tokens:
+- `"[PAD]"` - Padding token. Padding tokens are appended to the input sequence length
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:187]: Add line break before list

File: guides/keras_nlp/fnet_classification.py [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+input_sentence_ex = train_ds.take(1).get_single_element()[0][0]
+input_tokens_ex = tokenizer(input_sentence_ex)
+
+print("Sentence: ", input_sentence_ex)
+print("Tokens: ", input_tokens_ex)
+print("Recovered text after detokenizing: ", tokenizer.detokenize(input_tokens_ex))
+
+
+"""
+## Formatting the Dataset
```
Comment (fchollet) [inline guides/keras_nlp/fnet_classification.py:224]: Only capitalize the first word in a section title

**Review by fchollet — COMMENTED** [review]
Thanks!

**Review by abheesht17 — COMMENTED** [review]
Thanks for the review, @fchollet! Addressed your comments :)

File: examples/nlp/fnet_classification_with_keras_nlp.py [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+In 2020, a paper titled
+[FNet: Mixing Tokens with Fourier Transforms](https://arxiv.org/abs/2105.03824)
+replaced the self-attention layer in BERT with a simple Fourier Transform layer
+for "token mixing". This resulted in comparable accuracy and a speed-up during
+training. In particular, a couple of points from the paper stand out:
+
+1. The authors claim that FNet is 80% faster than BERT on GPUs and 70% faster
+on TPUs.
+The reason for this speed-up is two-fold:
+    a. The Fourier Transform layer is unparametrized, it does not have any parameters!
```
Comment (fchollet) [inline examples/nlp/fnet_classification_with_keras_nlp.py:41]: Does this get rendered as a nested list when you generate the website?

**Review by fchollet — COMMENTED** [review]
LGTM, thanks for the updates!

**Review by mattdangerw — APPROVED** [review]

**Review by fchollet — APPROVED** [review]
Thank you for the great contribution! 👍

**Comment by abheesht17** [comment]
Thank you, @mattdangerw, @fchollet for the review comments and the approval! :)

**Comment by fchollet** [comment]
> Continuous integration / black (pull_request) Failing after 23s — black

Please fix the code formatting.

**Comment by fchollet** [comment]
The CI error is unrelated. I'll merge. Thank you!
